toolchain/semantics/semantics_ir_formatter.cpp

```diff
@@ -0,0 +1,654 @@
+  auto GetLabelFor(ScopeIndex scope_idx, SemanticsNodeBlockId block_id)
+      -> std::string {
+    if (!block_id.is_valid()) {
+      return "!invalid";
+    }
+
+    auto& [label_scope, label_name] = labels[block_id.index];
+    if (!label_name) {
+      // This should not happen in valid IR.
+      return "<unexpected nodeblockref " + llvm::itostr(block_id.index) + ">";
+    }
+    if (label_scope == scope_idx) {
+      return label_name;
+    }
+    return (GetScopeInfo(label_scope).name.str() + "." + label_name.str())
+        .str();
+  }
+
+ private:
+  // A space in which unique names can be allocated.
+  struct Namespace {
+    // A result of a name lookup.
+    struct NameResult;
+
+    // A name in a namespace, which might be redirected to refer to another name
+    // for disambiguation purposes.
+    class Name {
+     public:
+      Name() : value_(nullptr) {}
+      explicit Name(llvm::StringMapIterator<NameResult> it) : value_(&*it) {}
+
+      explicit operator bool() const { return value_; }
+
+      auto str() const -> llvm::StringRef {
+        llvm::StringMapEntry<NameResult>* value = value_;
+        CARBON_CHECK(value) << "cannot print a null name";
+        while (value->second.ambiguous && value->second.fallback) {
+          value = value->second.fallback.value_;
+        }
+        return value->first();
+      }
+
+      operator llvm::StringRef() const { return str(); }
+      operator llvm::Twine() const { return str(); }
+      operator std::string() const { return str().str(); }
+
+      auto SetFallback(Name name) -> void { value_->second.fallback = name; }
+
+      auto SetAmbiguous() -> void { value_->second.ambiguous = true; }
+
+     private:
+      llvm::StringMapEntry<NameResult>* value_;
+    };
+
+    struct NameResult {
+      bool ambiguous = false;
+      Name fallback = Name();
+    };
+
+    llvm::StringRef prefix;
+    llvm::StringMap<NameResult> allocated = {};
+    int unnamed_count = 0;
+
+    auto AddNameUnchecked(llvm::StringRef name) -> Name {
+      return Name(allocated.insert({name, NameResult()}).first);
+    }
+
+    auto AllocateName(const NodeNamer& namer, ParseTree::Node node,
+                      std::string name = "") -> Name {
+      // The best (shortest) name for this node so far, and the current name
+      // for it.
+      Name best;
+      Name current;
+
+      // Add `name` as a name for this entity.
+      auto add_name = [&](bool mark_ambiguous = true) {
+        auto [it, added] = allocated.insert({name, NameResult()});
+        Name new_name = Name(it);
+
+        if (!added) {
+          if (mark_ambiguous) {
+            // This name was allocated for a different node. Mark it as
+            // ambiguous and keep looking for a name for this node.
+            new_name.SetAmbiguous();
+          }
+        } else {
+          if (!best) {
+            best = new_name;
+          } else {
+            current.SetFallback(new_name);
```

Comment:
To get a slightly easier to debug crash:

```suggestion
          } else {
            CARBON_CHECK(current);
            current.SetFallback(new_name);
```